### PR TITLE
Removing numpy requirement from all files in examples/pytorch/domain_templates 

### LINKED ISSUE #16649: Drop the numpy requirement
### Outline & Motivation

We barely use `numpy`. We should replace its uses with torch.

Reduce the package footprint

### Pitch

Remove https://github.com/Lightning-AI/lightning/blob/7bbbe22636853e1ee6fa57b557b3408d45233589/requirements/pytorch/base.txt#L4

Update numpy usages with torch.

### Additional context

_No response_

cc @borda @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by Bhavay-2001**
## What does this PR do?

Part of #16649 

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
@Borda @carmocca @awaelchli 
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19947.org.readthedocs.build/en/19947/

### CODE REVIEW

**Comment by Bhavay-2001** [comment]
Hi, I will be contributing and changing all of the files in the domain_templates folder.

**Review by awaelchli — APPROVED** [review]

File: examples/pytorch/domain_templates/generative_adversarial_net.py [hunk 1]
```diff
@@ -59,7 +58,7 @@ def block(in_feat, out_feat, normalize=True):
             *block(128, 256),
             *block(256, 512),
             *block(512, 1024),
-            nn.Linear(1024, int(np.prod(img_shape))),
+            nn.Linear(1024, int(torch.prod(img_shape))),
```
Comment (awaelchli) [inline examples/pytorch/domain_templates/generative_adversarial_net.py:61]: `torch.prod` only works on tensors, unlike the previous numpy call which works on iterables.
So we could use `math.prod` here as a replacement
```suggestion
            nn.Linear(1024, int(math.prod(img_shape))),
```
(and add missing math import)

File: examples/pytorch/domain_templates/generative_adversarial_net.py [hunk 2]
```diff
@@ -80,7 +79,7 @@ def __init__(self, img_shape):
         super().__init__()
 
         self.model = nn.Sequential(
-            nn.Linear(int(np.prod(img_shape)), 512),
+            nn.Linear(int(torch.prod(img_shape)), 512),
```
Comment (awaelchli) [inline examples/pytorch/domain_templates/generative_adversarial_net.py:82]: same here :)

**Comment by Bhavay-2001** [comment]
Hi @awaelchli, I was looking at this [line](https://github.com/Lightning-AI/pytorch-lightning/blob/812ffdec84189bdee560d4b629df01dd2a6bfb53/examples/pytorch/domain_templates/reinforce_learn_Qnet.py#L106) as it has numpy dependency. I tried to look for a torch alternative for this but there aren't any direct functions but some workarounds possible. Should I try to change it with that or are there any other libraries supporting this function?

**Comment by awaelchli** [comment]
The equivalent function from the [Python `random` standard library](https://docs.python.org/3/library/random.html#random.sample) would be

```py
random.sample(range(len(self.buffer)), batch_size)
```

Generally speaking, for examples it's fine to use additional dependencies if you can't replace them. The original issue https://github.com/Lightning-AI/pytorch-lightning/issues/16649 is more concerned with removing the dependency from the package.

**Comment by Bhavay-2001** [comment]
> The equivalent function from the [Python `random` standard library](https://docs.python.org/3/library/random.html#random.sample) would be
> 
> ```python
> random.sample(range(len(self.buffer)), batch_size)
> ```
> 
> Generally speaking, for examples it's fine to use additional dependencies. If you can't replace the numpy requirement somewhere, that's fine. The original issue #16649 is more concerned with removing the dependency from the package.

Once this gets completed, I will try to change `numpy` to `torch` requirements in `src/lightning` module as well. Thanks

**Comment by awaelchli** [comment]
Sounds good, thanks for the effort. 
@Peiffap was also working on that issue https://github.com/Lightning-AI/pytorch-lightning/issues/16649, you may want to coordinate in case both of you are planning to submit PRs :) Thanks

**Comment by Bhavay-2001** [comment]
Alright. I will connect with him in the main issue. Thanks
